### PR TITLE
128 scclib for javascript

### PR DESCRIPTION
**Pull request by mrlomar**
## Content
Client computers should be able to be written with a javascript library in NodeJS and Angular

## Features

Change directory structure cc_library
Added Javascript lib version
Added NodeJS example
Added Angular example
Published to [npm](https://www.npmjs.com/package/js-scc)

Closes GH-128

### CODE REVIEW

File: cc_library/js-scc/index.js [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    this.contents = contents;
+  }
+}
+
+/**
+ * Abstract device class from which all custom devices should inherit.
+ * Defines all required methods needer for communication to S.C.I.L.E.R.
+ * @Abstract
+ */
+class Device {
```
Comment (IssaHanou) [inline cc_library/js-scc/index.js]: this should be a separate class

File: cc_library/js-scc/index.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+  statusChanged() {
+    this._sendMessage(
+      "back-end",
+      new Message(this.name, "status", this.device.getStatus())
+    );
+  }
+}
+
+const formatDate = function(date) {
```
Comment (IssaHanou) [inline cc_library/js-scc/index.js]: documentation

File: cc_library/js-scc/index.js [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+      onFailure: () => {
+        this._onConnectFailure();
+      },
+      willMessage: will,
+      reconnect: true,
+      keepAliveInterval: 10
+    });
+  }
+
+  statusChanged() {
```
Comment (IssaHanou) [inline cc_library/js-scc/index.js]: documentation

File: cc_library/py-scc/src/sciler/scclib/app.py [hunk 4]
```diff
@@ -247,7 +246,9 @@ def __check_message(self, contents):
                 self.device.reset()
                 logging.info(("instruction performed", instruction))
             else:
-                (success, failed_action) = self.device.perform_instruction(action)
+                (success, failed_action) = self.device.perform_instruction(
+                    action
+                )  # TODO: remove failed_action as return argument
```
Comment (IssaHanou) [inline cc_library/py-scc/src/sciler/scclib/app.py]: is this something to do in this MR?

File: cc_library/py-scc/src/sciler/scclib/app.py [hunk 5]
```diff
@@ -15,12 +15,11 @@ def __init__(self, config, device):
         Initialize device with its configuration json file and python script.
         """
         self.device = device
-        self.config = json.load(config)
-        self.name = self.config.get("id")
-        self.info = self.config.get("description")
-        self.host = self.config.get("host")
-        self.port = self.config.get("port")
-        self.labels = self.config.get("labels")
+        self.name = config.get("id")
```
Comment (IssaHanou) [inline cc_library/py-scc/src/sciler/scclib/app.py]: why remove self?

File: cc_library/js-scc/index.js [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    this.device_id = device_id;
+    this.time_sent = formatDate(new Date());
+    this.type = type;
+    this.contents = contents;
+  }
+}
+
+/**
+ * Abstract device class from which all custom devices should inherit.
+ * Defines all required methods needer for communication to S.C.I.L.E.R.
```
Comment (GwennanS) [inline cc_library/js-scc/index.js]: typo needer

File: cc_library/js-scc/index.js (hunk 1 above)
Comment (mrlomar) [reply cc_library/js-scc/index.js]: it is a separate class :P
I shall put this class in a different file

File: cc_library/py-scc/src/sciler/scclib/app.py (hunk 5 above)
Comment (mrlomar) [reply cc_library/py-scc/src/sciler/scclib/app.py]: self.config wasn't used

File: cc_library/py-scc/src/sciler/scclib/app.py (hunk 4 above)
Comment (mrlomar) [reply cc_library/py-scc/src/sciler/scclib/app.py]: I want to keep the javascript version and the python as much a like as possible

File: cc_library/js-scc/index.js (hunk 2 above)
Comment (mrlomar) [reply cc_library/js-scc/index.js]: done

Comment (mrlomar) [reply cc_library/js-scc/index.js]: done

File: cc_library/js-scc/index.js (hunk 3 above)
Comment (mrlomar) [reply cc_library/js-scc/index.js]: done

File: cc_library/py-scc/src/sciler/scclib/app.py (hunk 5 above)
Comment (mrlomar) [reply cc_library/py-scc/src/sciler/scclib/app.py]: self.config wasn't used

File: cc_library/js-scc/index.js (hunk 6 above)
Comment (mrlomar) [reply cc_library/js-scc/index.js]: fixed

File: cc_library/example-scripts-js-scc/display-angular/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# Display
+
+This project was generated with [Angular CLI](https://github.com/angular/angular-cli) version 8.3.19.
```
[the remark this answers is not in the document]
Comment (mrlomar) [reply cc_library/example-scripts-js-scc/display-angular/README.md]: done

File: cc_library/example-scripts-js-scc/display-angular/src/app/display.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+          date.getHours() +
+          ":" +
+          date.getMinutes() +
+          ":" +
+          date.getSeconds()
+        );
+      };
+      console.log(
+        "time=" + formatDate(date) + " level=" + level + " msg=" + message
+      ); // call own logger
```
[the remark this answers is not in the document]
Comment (mrlomar) [reply cc_library/example-scripts-js-scc/display-angular/src/app/display.ts]: No, it's just a comment of what can be done here

File: cc_library/example-scripts-js-scc/display-angular/src/app/display.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
+const Device = require("../../../../js-scc"); // development
+// const SccLib = require("js-scc"); // production
```
[the remark this answers is not in the document]
Comment (mrlomar) [reply cc_library/example-scripts-js-scc/display-angular/src/app/display.ts]: done

File: cc_library/example-scripts-js-scc/display-angular/src/app/app.component.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import { Component, OnInit } from "@angular/core";
+import { Display } from "./display";
+import * as data from "../assets/display_config.json";
+const SccLib = require("../../../../js-scc"); // development
+// const SccLib = require("js-scc"); // production
```
[the remark this answers is not in the document]
Comment (mrlomar) [reply cc_library/example-scripts-js-scc/display-angular/src/app/app.component.ts]: done

**Comment by mrlomar** [comment]
Found a flaky test (not related to this issue)
I made a new issue for this GH-150

File: cc_library/py-scc/src/sciler/scclib/app.py (hunk 4 above)
Comment (mrlomar) [reply cc_library/py-scc/src/sciler/scclib/app.py]: I made an issue for this GH-149

File: cc_library/js-scc/index.js (hunk 1 above)
Comment (mrlomar) [reply cc_library/js-scc/index.js]: This costs me no too much time. 
I made an issue for this GH-151

File: cc_library/py-scc/src/sciler/scclib/app.py (hunk 5 above)
Comment (IssaHanou) [reply cc_library/py-scc/src/sciler/scclib/app.py]: config should be used for checking input and output types so it should still be a self variable

Comment (mrlomar) [reply cc_library/py-scc/src/sciler/scclib/app.py]: added it back

File: cc_library/py-scc/src/sciler/scclib/app.py [hunk 11]
```diff
@@ -15,12 +15,12 @@ def __init__(self, config, device):
         Initialize device with its configuration json file and python script.
         """
         self.device = device
-        self.config = json.load(config)
```
Comment (IssaHanou) [inline cc_library/py-scc/src/sciler/scclib/app.py]: can you change this back to original

Comment (mrlomar) [reply cc_library/py-scc/src/sciler/scclib/app.py]: your right, it should be json.load

**Pull request merged**